File: src/main/java/fr/openent/formulaire/service/ResponseFileService.java
package fr.openent.formulaire.service;

import fr.wseduc.webutils.Either;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

public interface ResponseFileService {
    /**
     * List all files of a specific response
     * @param responseId response identifier
     * @param handler function handler returning JsonArray data
     */
    void list(String responseId, Handler<Either<String, JsonArray>> handler);

    /**
     * List all files of a specific question
     * @param questionId question identifier
     * @param handler function handler returning JsonArray data
     */
    void listByQuestion(String questionId, Handler<Either<String, JsonArray>> handler);

    /**
     * List all files of a specific form
     * @param formId  form identifier
     * @param handler function handler returning JsonArray data
     */
    void listByForm(String formId, Handler<Either<String, JsonArray>> handler);

    /**
     * Get a specific file by id
     * @param fileId file identifier
     * @param handler function handler returning JsonObject data
     */
    void get(String fileId, Handler<Either<String, JsonObject>> handler);

    /**
     * Create a response_file
     * @param responseId response identifier
     * @param fileId file identifier
     * @param filename file name
     * @param type file type
     * @param handler function handler returning JsonObject data
     */
    void create(String responseId, String fileId, String filename, String type, Handler<Either<String, JsonObject>> handler);

    /**
     * Delete all response_files of specific responses
     * @param responseIds response identifiers
     * @param handler function handler returning JsonArray data
     */
    void deleteAllByResponse(JsonArray responseIds, Handler<Either<String, JsonArray>> handler);

    /**
     * Delete all response_files by their ids
     * @param fileIds responseFiles identifier
     * @param handler function handler returning JsonArray data
     */
    void deleteAll(JsonArray fileIds, Handler<Either<String, JsonArray>> handler);
}

File: common/src/main/resources/ts/services/ResponseFileService.ts
import {idiom, ng, notify} from 'entcore';
import http from 'axios';
import {DataUtils} from "../utils";

export interface ResponseFileService {
    list(responseId: number) : Promise<any>;
    listByQuestion(questionId: number) : Promise<any>;
    get(fileId: number) : Promise<any>;
    download(fileId: number) : Promise<any>;
    zipAndDownload(questionId: number) : Promise<any>;
    create(responseId: number, file) : Promise<any>;
    deleteAll(responseId : number) : Promise<any>;
    deleteAllMultiple(responseIds: number[]) : Promise<void>;
}

export const responseFileService: ResponseFileService = {

    async list(responseId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.get(`/formulaire/responses/${responseId}/files/all`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseFileService.list'));
            throw err;
        }
    },

    async listByQuestion(questionId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.get(`/formulaire/questions/${questionId}/files/all`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseFileService.list'));
            throw err;
        }
    },

    async get(fileId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.get(`/formulaire/responses/files/${fileId}`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseFileService.get'));
            throw err;
        }
    },

    async download(fileId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.get(`/formulaire/responses/files/${fileId}/download`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseFileService.download'));
            throw err;
        }
    },

    async zipAndDownload(questionId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.get(`/formulaire/responses/${questionId}/files/download/zip`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseFileService.download'));
            throw err;
        }
    },

    async create(responseId: number, file) : Promise<any> {
        try {
            return DataUtils.getData(await http.post(`/formulaire/responses/${responseId}/files`, file, {'headers': {'Content-Type': 'multipart/form-data'}}));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseFileService.create'));
            throw err;
        }
    },

    async deleteAll(responseId: number) : Promise<any> {
        try {
            return DataUtils.getData(await http.delete(`/formulaire/responses/${responseId}/files`));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseFileService.delete'));
            throw err;
        }
    },

    async deleteAllMultiple(responseIds: number[]) : Promise<void> {
        try {
            return DataUtils.getData(await http.delete(`/formulaire/responses/files/multiple`, { data: responseIds }));
        } catch (err) {
            notify.error(idiom.translate('formulaire.error.responseFileService.delete'));
            throw err;
        }
    }
};

export const ResponseFileService = ng.service('ResponseFileService', (): ResponseFileService => responseFileService);